### ISSUE TITLE
Crash when using Update From Tag function

### ISSUE THREAD

**Issue by sandrotosi** [puddletag/puddletag#524]
<a href="https://github.com/blueblots"><img src="https://avatars2.githubusercontent.com/u/63152708?v=4" align="left" width="96" height="96" hspace="10"></img></a> **Issue by [blueblots](https://github.com/blueblots)**
_Fri Jul 31 13:10:32 2020_
_Originally opened as https://github.com/sandrotosi/puddletag/issues/53_

----

Open puddletag, select some non-MP3 files and try to use the Update from Tag function with an ID3 tag: (the files I used for this are OGG format)

![updatefromtag-crash (1)](https://user-images.githubusercontent.com/63152708/89037933-502cd980-d337-11ea-8aef-0f9624de2411.png)

Puddletag crashes with this exception:

```
Traceback (most recent call last):
  File "../puddletag/source/puddlestuff/actiondlg.py", line 364, in showexample
    val = apply_actions([self.func], audio, state, fields)
  File "../puddletag/source/puddlestuff/findfunc.py", line 623, in apply_actions
    temp = func.runFunction(val, audio, state, None, r_tags)
  File "../puddletag/source/puddlestuff/findfunc.py", line 873, in runFunction
    return func(**topass)
  File "../puddletag/source/puddlestuff/functions.py", line 1037, in update_from_tag
    tag = _update[tag]().link(r_tags.filepath)
  File "../puddletag/source/puddlestuff/audioinfo/id3.py", line 1002, in link
    tags, audio = self.load(filename, id3_filetype)
  File "../puddletag/source/puddlestuff/audioinfo/util.py", line 838, in load
    audio = filetype(filename)
  File "/usr/lib/python3.8/site-packages/mutagen/_file.py", line 49, in __init__
    self.load(*args, **kwargs)
  File "../puddletag/source/puddlestuff/audioinfo/id3.py", line 761, in load
    return super(ID3FileType, self).load(filename, ID3, **kwargs)
  File "/usr/lib/python3.8/site-packages/mutagen/_util.py", line 140, in wrapper
    return func(self, h, *args, **kwargs)
  File "/usr/lib/python3.8/site-packages/mutagen/id3/_file.py", line 423, in load
    self.info = self._Info(fileobj, offset)
  File "/usr/lib/python3.8/site-packages/mutagen/_util.py", line 169, in wrapper
    return func(*args, **kwargs)
  File "/usr/lib/python3.8/site-packages/mutagen/mp3/__init__.py", line 403, in __init__
    raise HeaderNotFoundError("can't sync to MPEG frame")
mutagen.mp3.HeaderNotFoundError: can't sync to MPEG frame
../puddletag/launch.bash: line 3: 25495 Aborted                 (core dumped) PYTHONPATH=source/ ./source/puddletag
```

I've fixed the crash in https://github.com/sandrotosi/puddletag/pull/56, but it would be better if the function stopped immediately when it notices that there is no MP3 tags in the file and displayed a message to notify the user. Currently it just tries to continue doing`update_from_tag`, but only gets None values and empty dicts until it finishes.

This is a pdb trace of what happens when the `update_from_tag` function hits the mutagen error (with the fix applied so it does not core dump):

```
> ../puddletag/source/puddlestuff/functions.py(1045)update_from_tag()
-> return
(Pdb) next
--Return--
> ../puddletag/source/puddlestuff/functions.py(1045)update_from_tag()->None
-> return
(Pdb) next
--Return--
> ../puddletag/source/puddlestuff/findfunc.py(873)runFunction()->None
-> return func(**topass)
(Pdb) next
> ../puddletag/source/puddlestuff/findfunc.py(624)apply_actions()
-> if temp is None:
(Pdb) temp
(Pdb) next
> ../puddletag/source/puddlestuff/findfunc.py(625)apply_actions()
-> continue
(Pdb) l
620  
621             for field in fields:
622                 val = audio.get(field, '')
623                 temp = func.runFunction(val, audio, state, None, r_tags)
624                 if temp is None:
625  ->                 continue
626                 if isinstance(temp, str):
627                     ret[field] = temp
628                 elif hasattr(temp, 'items'):
629                     ret.update(temp)
630                     break
(Pdb) next
> ../puddletag/source/puddlestuff/findfunc.py(621)apply_actions()
-> for field in fields:
(Pdb) l
616                 fields = parse_field_list(func.tag, audio)
617             else:
618                 fields = ovr_fields[::]
619             ret = {}
620  
621  ->         for field in fields:
622                 val = audio.get(field, '')
623                 temp = func.runFunction(val, audio, state, None, r_tags)
624                 if temp is None:
625                     continue
626                 if isinstance(temp, str):
(Pdb) field
'just nothing to do with this'
(Pdb) fields
['just nothing to do with this']
(Pdb) next
> ../puddletag/source/puddlestuff/findfunc.py(643)apply_actions()
-> ret = dict([z for z in ret.items() if z[1] is not None])
(Pdb) l
638                         ret[field] = temp[0]
639                     else:
640                         ret[field] = temp
641                 else:
642                     ret[field] = temp[0]
643  ->         ret = dict([z for z in ret.items() if z[1] is not None])
644             if ret:
645                 [changed.add(z) for z in ret]
646                 audio.update(ret)
647         return dict([(z, audio[z]) for z in changed])
648  
(Pdb) ret
{}
(Pdb) next
> ../puddletag/source/puddlestuff/findfunc.py(644)apply_actions()
-> if ret:
(Pdb) ret
{}
(Pdb) next
> ../puddletag/source/puddlestuff/findfunc.py(614)apply_actions()
-> for func in actions:
(Pdb) ret
{}
(Pdb) func
<puddlestuff.findfunc.Function object at 0x7f0b7c03d580>
(Pdb) actions
[<puddlestuff.findfunc.Function object at 0x7f0b7c03d580>]
(Pdb) actions.function
*** AttributeError: 'list' object has no attribute 'function'
(Pdb) actions.info
*** AttributeError: 'list' object has no attribute 'info'
(Pdb) actions.doc
*** AttributeError: 'list' object has no attribute 'doc'
(Pdb) next
> ../puddletag/source/puddlestuff/findfunc.py(647)apply_actions()
-> return dict([(z, audio[z]) for z in changed])
(Pdb) l
642                     ret[field] = temp[0]
643             ret = dict([z for z in ret.items() if z[1] is not None])
644             if ret:
645                 [changed.add(z) for z in ret]
646                 audio.update(ret)
647  ->     return dict([(z, audio[z]) for z in changed])
648  
649  
650     def apply_macros(macros, audio, state, fields=None):
651         actions = []
652         for m in macros:
(Pdb) l
653             actions.extend(m.actions)
654  
655         if fields:
656             return apply_actions(actions, audio, state, fields)
657         else:
658             return apply_actions(actions, audio, state)
659  
660  
661     def runQuickAction(funcs, audio, state, tag):
662         """Same as runAction, except that all funcs are
663         applied not in the values stored but on audio[tag]."""
(Pdb) next
--Return--
> ../puddletag/source/puddlestuff/findfunc.py(647)apply_actions()->{}
-> return dict([(z, audio[z]) for z in changed])
(Pdb) z
*** NameError: name 'z' is not defined
(Pdb) changed
set()
(Pdb) next
> ../puddletag/source/puddlestuff/actiondlg.py(367)showexample()
-> if val is not None:
(Pdb) val
{}
(Pdb) next
> ../puddletag/source/puddlestuff/actiondlg.py(368)showexample()
-> self.updateExample.emit(val)
(Pdb) l
363                     state = {'__counter': '0', '__total_files': files}
364                     val = apply_actions([self.func], audio, state, fields)
365                 except findfunc.ParseError as e:
366                     val = '<b>%s</b>' % (e.message)
367                 if val is not None:
368  ->                 self.updateExample.emit(val)
369                 else:
370                     self.updateExample.emit(translate('Functions Dialog', '<b>No change</b>'))
371  
372         def _sanitize(self, ctype, value):
373             if ctype in ['combo', 'text']:
(Pdb) next
--Return--
> ../puddletag/source/puddlestuff/actiondlg.py(368)showexample()->None
-> self.updateExample.emit(val)
(Pdb) l
363                     state = {'__counter': '0', '__total_files': files}
364                     val = apply_actions([self.func], audio, state, fields)
365                 except findfunc.ParseError as e:
366                     val = '<b>%s</b>' % (e.message)
367                 if val is not None:
368  ->                 self.updateExample.emit(val)
369                 else:
370                     self.updateExample.emit(translate('Functions Dialog', '<b>No change</b>'))
371  
372         def _sanitize(self, ctype, value):
373             if ctype in ['combo', 'text']:
(Pdb) next
> ../puddletag/source/puddlestuff/actiondlg.py(404)setArguments()
-> for argument, control in zip(args, self.controls):
(Pdb) next
--Return--
> ../puddletag/source/puddlestuff/actiondlg.py(404)setArguments()->None
-> for argument, control in zip(args, self.controls):
(Pdb) next
--Return--
> ../puddletag/source/puddlestuff/actiondlg.py(335)loadSettings()->None
-> self.setArguments(fields, arguments)
(Pdb) next
> ../puddletag/source/puddlestuff/actiondlg.py(496)createWindow()
-> self.stackWidgets.update({index: widget})
(Pdb) next
> ../puddletag/source/puddlestuff/actiondlg.py(497)createWindow()
-> self.stack.addWidget(widget)
(Pdb) next
> ../puddletag/source/puddlestuff/actiondlg.py(498)createWindow()
-> widget.updateExample.connect(self.updateExample)
(Pdb) next
> ../puddletag/source/puddlestuff/actiondlg.py(499)createWindow()
-> self.stack.setCurrentWidget(self.stackWidgets[index])
(Pdb) next
> ../puddletag/source/puddlestuff/actiondlg.py(500)createWindow()
-> self.stackWidgets[index].showexample()
(Pdb) next
> ../puddletag/source/puddlestuff/functions.py(1045)update_from_tag()
-> return
(Pdb) next
--Return--
> ../puddletag/source/puddlestuff/functions.py(1045)update_from_tag()->None
-> return
(Pdb) next
--Return--

```

**Comment by blueblots** [comment]
This was fixed in #543, and should be closed now.

**Comment by sandrotosi** [comment]
Fixed via #543
